### ISSUE TITLE
Reduce number ClassEnv_classHasIllegalStaticFinalFieldModification messages

### ISSUE THREAD

**Issue by mpirvu** [eclipse/openj9#9358]
A recent openj9 fix (#9246) has increased the number of `ClassEnv_classHasIllegalStaticFinalFieldModification` queries the JITServer sends to the client, from 0 to about 8K for AcmeAir). If possible we would like to reduce the number of such messages.
The frontEnd implementation for non-jitserver is
```
J9::ClassEnv::classHasIllegalStaticFinalFieldModification(TR_OpaqueClassBlock * clazzPointer)
   {
   J9Class* j9clazz = TR::Compiler->cls.convertClassOffsetToClassPtr(clazzPointer);
   return J9_ARE_ANY_BITS_SET(j9clazz->classFlags, J9ClassHasIllegalFinalFieldModifications);
   }
```
So, the query only looks at the `J9ClassHasIllegalFinalFieldModifications` bit inside `j9class->classFlags`.
This bit is set by the `jitIllegalFinalFieldModification(J9VMThread *currentThread, J9Class *fieldClass)` hook and never reset.
Because the setting can happen at any time, caching a value of 0 will not work. We can use caching after the bit has been set, but I am afraid that in real life this bit is set very infrequently and thus, our caching mechanism is not going to be effective at all.

The `jitIllegalFinalFieldModification` hook, besides setting the aforementioned flag will also use the runtimeAssumptionTable to do any required patching.
The corresponding assumptions are set in `commitOSRVirtualGuards()` and `commitVirtualGuard()` at the time of CHTable commit which is done at the end of the compilation. However, it's not clear to me how a compilation should behave when finalFieldModification happens during a compilation, but after the test has been performed. For example:
1) Check `J9ClassHasIllegalFinalFieldModifications` bit which is clear
2) Perform the static field folding
3) Final field is changed during compilation and `J9ClassHasIllegalFinalFieldModifications` is set. Assumption patching is triggered, but at this point there is no assumption set yet
4) CHTable->commit() registers the assumption()

**Comment by mpirvu** [comment]
I went to the original implementation (PR #2580) and was able to infer the following:
The compiler object maintains a list (TR_Array) of classes for which their final field was folded during the current compilation. At the end of the compilation, during CHTable.commit() we go over this list and if any class has the `J9ClassHasIllegalFinalFieldModifications` bit set we fail the compilation with an appropriate error code.
I thought there was a synchronization hole, but during the CHTable.commit() we have VMAcess while the VM will have stopped all threads when setting `J9ClassHasIllegalFinalFieldModifications` (with `vmFuncs->acquireSafePointVMAccess(currentThread)`, so there is race condition.

**Comment by mpirvu** [comment]
The solution to reducing the number of `ClassEnv_classHasIllegalStaticFinalFieldModification` could be the following:
Maintain a temporary list of classes that have the `J9ClassHasIllegalFinalFieldModifications` bit set. This is very similar to the temporary list of unloaded classes: `compInfo->getUnloadedClassesTempList()`. Every time the `jitIllegalFinalFieldModification()` hook is called for a class, add the respective class to this temporary list. 
When a compilation request is sent to JITServer, take all the classes from this temporary list and send them to JITServer. Clear this temporary list (again, very similar to what we do for unloaded classes):
```
  compInfo->getSequencingMonitor()->enter();
   // Collect the list of unloaded classes
   std::vector<TR_OpaqueClassBlock*> unloadedClasses(compInfo->getUnloadedClassesTempList()->begin(), compInfo->getUnloadedClassesTempList()->end());
   compInfo->getUnloadedClassesTempList()->clear();
   // Collect and encode the CHTable updates; this will acquire CHTable mutex
   auto table = (JITClientPersistentCHTable*)compInfo->getPersistentInfo()->getPersistentCHTable();
   std::pair<std::string, std::string> chtableUpdates = table->serializeUpdates();
   // Update the sequence number for these updates
   uint32_t seqNo = compInfo->getCompReqSeqNo();
   compInfo->incCompReqSeqNo();
   compInfo->getSequencingMonitor()->exit();
```
JITServer will process this `IllegalStaticFinalFieldModificationList` at the same time when the unloaded class are processed with `clientSession->processUnloadedClasses()`
Here, we just add the `J9ClassHasIllegalFinalFieldModifications` bit to our cached classes.

It's possible that a compilation will not see the `J9ClassHasIllegalFinalFieldModifications` bit being set in time, but that is not a functional issue: during CHTable.commit we will check the list of classes which took advantage of final field folding and fail this compilation if any class in that list has the `J9ClassHasIllegalFinalFieldModifications` bit set. Thus, what we have here is a larger window of time where updates from the VM are not visible to the server and thus a larger likelihood to fail compilations. However, these events are rare in nature and I don't expect any performance delta.

**Comment by a7ehuo** [comment]
I'll look at this one next.

**Comment by a7ehuo** [comment]
> what we have here is a larger window of time where updates from the VM are not visible to the server and thus a larger likelihood to fail compilations. 

Just thinking, as long as the `J9ClassHasIllegalFinalFieldModifications` is set,  the compilation will be aborted in the end when committing CHTable. When Optimizer finds that the bit `J9ClassHasIllegalFinalFieldModifications` is set, I wonder if it should abort the compilation already instead of carrying on the compilation.

**Comment by mpirvu** [comment]
> as long as the J9ClassHasIllegalFinalFieldModifications is set, the compilation will be aborted in the end when committing CHTable

No, if the JITServer/optimizer sees the  `J9ClassHasIllegalFinalFieldModifications` bit being set, it will not perform the static field folding.
The compilation is aborted only if the optimizer didn't see the bit being set, performed static field folding, but during CHTable commit realized it shouldn't have because the bit was actually set.

**Comment by mpirvu** [comment]
Issue fixed by #9415
